substratevm/src/com.oracle.svm.core/src/com/oracle/svm/core/c/libc/LibCBase.java

```diff
@@ -24,18 +24,53 @@
  */
 package com.oracle.svm.core.c.libc;
 
+import java.lang.reflect.AnnotatedElement;
+import java.lang.reflect.Method;
 import java.nio.file.Path;
+import java.util.Arrays;
 import java.util.List;
 
+import org.graalvm.nativeimage.ImageSingletons;
+
 public interface LibCBase {
 
     String PATH_DEFAULT = "<default>";
 
+    static <T extends AnnotatedElement> boolean isProvidedInCurrentLibc(T element) {
+        LibCBase currentLibC = ImageSingletons.lookup(LibCBase.class);
+        Libc targetLibC = element.getAnnotation(Libc.class);
+        return targetLibC == null || Arrays.asList(targetLibC.value()).contains(currentLibC.getClass());
+    }
+
+    static boolean isMethodProvidedInCurrentLibc(Method method) {
+        if (!isProvidedInCurrentLibc(method)) {
+            return false;
+        }
+        Class<?> declaringClass = method.getDeclaringClass();
+        return isProvidedInCurrentLibc(declaringClass);
+    }
+
+    static boolean isTypeProvidedInCurrentLibc(Class<?> clazz) {
+        if (!isProvidedInCurrentLibc(clazz)) {
+            return false;
+        }
+        Class<?> enclosingClass = clazz.getEnclosingClass();
+        return enclosingClass == null || isProvidedInCurrentLibc(enclosingClass);
```

Comment:
This should work with arbitrarily nested classes/packages. See `com.oracle.svm.hosted.ImageClassLoader#handleClass` for reference.

Reply:
I agree, I have modified the code accordingly.